hackerpet/__init__.py

```diff
@@ -0,0 +1,494 @@
+    """
+
+    STANDBY = "Standby"
+    ACTIVE = "Active"
+
+    @classmethod
+    def _missing_(cls: type, value: Any) -> str:
+        raise IllegalValue(cls, value)
+
+
+class MaxKibbles:
+    """
+    a limit on the number of kibbles dispensed in a day.
+
+    If limit == 0 or None, disables limit
+    """
+
+    _limit: Optional[int]
+    value: int
+
+    def __init__(self, limit: Optional[int]):
+        self._limit = limit
+        self.value = limit if limit else 0
+
+    def __str__(self) -> str:
+        if self._limit == 0 or self._limit is None:
+            return "no limit"
+        else:
+            return f"limit: {self._limit}"
+
+    def __repr__(self) -> str:
+        return f"<{type(self).__name__}: {self.value}>"
+
+
+class Schedule:
+    """
+    defines the schedule that the hackerpet will follow in "Scheduled" HubMode.
+
+    Sets the schedule for weekdays and weekends separately
+    """
+
+    weekday_from: time
+    weekday_to: time
+    weekend_from: time
+    weekend_to: time
+
+    def __init__(
+        self, weekday_from: time, weekday_to: time, weekend_from: time, weekend_to: time
+    ):
+        self.weekday_from = weekday_from
+        self.weekday_to = weekday_to
+        self.weekend_from = weekend_from
+        self.weekend_to = weekend_to
+
+    @staticmethod
+    def _hm_format(tme: time) -> str:
+        return tme.strftime("%H:%S")
+
+    def __str__(self) -> str:
+        return (
+            "schedule: weekdays "
+            f"{Schedule._hm_format(self.weekday_from)} - "
+            f"{Schedule._hm_format(self.weekday_to)}; "
+            f"weekends {Schedule._hm_format(self.weekend_from)} - "
+            f"{Schedule._hm_format(self.weekend_to)}"
+        )
+
+    def __repr__(self) -> str:
+        return (
+            f"{type(self).__name__}: "
+            f"<weekday_from: {self.weekday_from.__repr__()}, "
+            f"weekday_to: {self.weekday_to.__repr__()}, "
+            f"weekend_from: {self.weekend_from.__repr__()}, "
+            f"weekend_to: {self.weekday_to.__repr__()}>"
+        )
+
+    @classmethod
+    def from_json(cls: type, payload: Any) -> "Schedule":
+        """parse json response to instantiate Schedule"""
+        weekday_from = payload["weekday_from"]
+        weekday_to = payload["weekday_to"]
+        weekend_from = payload["weekend_from"]
+        weekend_to = payload["weekend_to"]
+        return cls(weekday_from, weekday_to, weekend_from, weekend_to)
+
+    def payload(self) -> str:
+        """hackerpet expects this exact layout in json payload"""
+
+        return json.dumps(
+            {
+                "weekday_from": Schedule._hm_format(self.weekday_from),
+                "weekday_to": Schedule._hm_format(self.weekday_to),
+                "weekend_from": Schedule._hm_format(self.weekend_from),
+                "weekend_to": Schedule._hm_format(self.weekend_to),
+            },
+            separators=(",", ":"),
+        )
+
+
+class Status:
```

Comment:
What would you think about breaking this out into a Status dataclass and a constructor that created it from the raw body? I understand that this is a recommended pattern by HomeAssistant, but not sure why it is a beneficial pattern.

Reply:
tried it. think it's better